File: android/calliopeia-sdk/src/main/kotlin/com/calliopeia/sdk/CalliopeiaModels.kt
package com.calliopeia.sdk

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import java.net.URI
import java.util.UUID

enum class CalliopeiaCredentialType(val apiValue: String) {
    JWT("JWT"),
    API_KEY("API_KEY"),
}

data class CalliopeiaCredential(
    val value: String,
    val type: CalliopeiaCredentialType = CalliopeiaCredentialType.JWT,
)

fun interface CalliopeiaCredentialProvider {
    suspend fun credential(): CalliopeiaCredential
}

class StaticCalliopeiaCredentialProvider(
    private val storedCredential: CalliopeiaCredential,
) : CalliopeiaCredentialProvider {
    override suspend fun credential(): CalliopeiaCredential = storedCredential
}

data class CalliopeiaAPIConfiguration(
    val graphQLEndpoint: URI,
    val appSyncAPIKey: String,
    val pullAPIBaseURL: URI? = null,
)

enum class CalliopeiaResponseMode(val apiValue: String) {
    ASYNC("ASYNC"),
    SUBSCRIPTION("SUBSCRIPTION"),
    WEBHOOK("WEBHOOK"),
}

enum class CalliopeiaExtractionEffort(val apiValue: String) {
    LOW("LOW"),
    STANDARD("STANDARD"),
    MAXIMUM("MAX"),
}

enum class CalliopeiaAuditMode(val apiValue: String) {
    OFF("OFF"),
    OBSERVE("OBSERVE"),
    ENFORCE("ENFORCE"),
}

enum class CalliopeiaAuditEffort(val apiValue: String) {
    STANDARD("STANDARD"),
    MAXIMUM("MAX"),
}

enum class CalliopeiaAuditStrategy(val apiValue: String) {
    COMBINED("COMBINED"),
    ATOMIC_BATCH("ATOMIC_BATCH"),
}

data class CalliopeiaAudioJobRequest(
    val idempotencyKey: String = UUID.randomUUID().toString(),
    val processingProfileID: String? = null,
    val extractionEffort: CalliopeiaExtractionEffort? = null,
    val auditMode: CalliopeiaAuditMode? = null,
    val auditEffort: CalliopeiaAuditEffort? = null,
    val auditStrategy: CalliopeiaAuditStrategy? = null,
    val auditBatchSize: Int? = null,
    val promptText: String? = null,
    val promptTemplateID: String? = null,
    val promptTitle: String? = null,
    val responseMode: CalliopeiaResponseMode = CalliopeiaResponseMode.ASYNC,
    val webhookEndpointID: String? = null,
    val userID: String? = null,
    val shopID: String? = null,
    val customerID: String? = null,
    val expectedAddressee: String? = null,
    val karteID: String? = null,
    val summaryID: String? = null,
    val passthrough: JsonObject? = null,
)

data class CalliopeiaUploadTicket(
    val objectKey: String,
    val uploadURL: URI,
    val method: String,
    val contentType: String,
    val expiresAt: String?,
)

data class CalliopeiaAcceptedJob(
    val id: String,
    val status: String,
    val externalSummaryID: String?,
    val externalIdempotencyKey: String?,
    val externalResponseMode: String?,
    val externalCallbackStatus: String?,
)

data class CalliopeiaJobSubmission(
    val message: String?,
    val idempotentReplay: Boolean,
    val subscriptionToken: String?,
    val statusURL: URI?,
    val job: CalliopeiaAcceptedJob,
)

data class CalliopeiaJobSnapshot(
    val id: String,
    val status: String,
    val operation: String?,
    val inputKind: String?,
    val fileName: String?,
    val audioSeconds: Double?,
    val responseText: String?,
    val responseJSON: JsonElement?,
    val extractionEffort: String?,
    val auditMode: String?,
    val auditEffort: String?,
    val auditStrategy: String?,
    val auditBatchSize: Int?,
    val auditState: String?,
    val auditJSON: JsonElement?,
    val deliveryBlockedReason: String?,
    val errorMessage: String?,
    val costUSD: Double?,
    val costJPY: Double?,
    val createdAt: String?,
    val updatedAt: String?,
    val completedAt: String?,
    val externalShopID: String?,
    val externalCustomerID: String?,
    val externalKarteID: String?,
    val externalSummaryID: String?,
    val passthrough: JsonElement?,
)

data class CalliopeiaPullJob(
    val version: String,
    val jobID: String,
    val status: String,
    val shopID: String?,
    val customerID: String?,
    val karteID: String?,
    val summaryID: String?,
    val passthrough: JsonElement?,
    val error: JsonElement?,
    val delivery: JsonElement,
    val createdAt: String?,
    val updatedAt: String?,
    val completedAt: String?,
    val result: JsonElement?,
)

data class CalliopeiaPullJobResponse(
    val success: Boolean,
    val job: CalliopeiaPullJob,
    val requestID: String,
)

sealed class CalliopeiaSDKException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class InvalidRequest(message: String) : CalliopeiaSDKException(message)
    class InvalidResponse(message: String, cause: Throwable? = null) : CalliopeiaSDKException(message, cause)
    class Service(message: String) : CalliopeiaSDKException(message)
    class HTTP(val statusCode: Int, val responseBody: String?) : CalliopeiaSDKException(
        responseBody?.let { "Calliopeia returned HTTP $statusCode: $it" }
            ?: "Calliopeia returned HTTP $statusCode",
    )
    class MissingPullAPIBaseURL : CalliopeiaSDKException(
        "pullAPIBaseURL is required to retrieve job results",
    )
}

File: ios/Sources/CalliopeiaSDK/CalliopeiaModels.swift
import Foundation

public enum CalliopeiaCredentialType: String, Codable, Sendable {
    case jwt = "JWT"
    case apiKey = "API_KEY"
}

public struct CalliopeiaCredential: Sendable {
    public let value: String
    public let type: CalliopeiaCredentialType

    public init(value: String, type: CalliopeiaCredentialType = .jwt) {
        self.value = value
        self.type = type
    }
}

public protocol CalliopeiaCredentialProvider: Sendable {
    func credential() async throws -> CalliopeiaCredential
}

public struct ClosureCalliopeiaCredentialProvider: CalliopeiaCredentialProvider {
    public typealias Loader = @Sendable () async throws -> CalliopeiaCredential

    private let loader: Loader

    public init(loader: @escaping Loader) {
        self.loader = loader
    }

    public func credential() async throws -> CalliopeiaCredential {
        try await loader()
    }
}

public struct StaticCalliopeiaCredentialProvider: CalliopeiaCredentialProvider {
    private let storedCredential: CalliopeiaCredential

    public init(_ credential: CalliopeiaCredential) {
        self.storedCredential = credential
    }

    public func credential() async throws -> CalliopeiaCredential {
        storedCredential
    }
}

public enum CalliopeiaGraphQLAuthorization: Sendable {
    case apiKey
    case cognitoUserPools
}

public struct CalliopeiaAPIConfiguration: Sendable {
    public let graphQLEndpoint: URL
    public let appSyncAPIKey: String
    public let pullAPIBaseURL: URL?
    public let graphQLAuthorization: CalliopeiaGraphQLAuthorization

    public init(
        graphQLEndpoint: URL,
        appSyncAPIKey: String,
        pullAPIBaseURL: URL? = nil,
        graphQLAuthorization: CalliopeiaGraphQLAuthorization = .apiKey
    ) {
        self.graphQLEndpoint = graphQLEndpoint
        self.appSyncAPIKey = appSyncAPIKey
        self.pullAPIBaseURL = pullAPIBaseURL
        self.graphQLAuthorization = graphQLAuthorization
    }
}

public enum CalliopeiaResponseMode: String, Codable, Sendable {
    case async = "ASYNC"
    case subscription = "SUBSCRIPTION"
    case webhook = "WEBHOOK"
}

public enum CalliopeiaExtractionEffort: String, Codable, Sendable {
    case low = "LOW"
    case standard = "STANDARD"
    case maximum = "MAX"
}

public enum CalliopeiaAuditMode: String, Codable, Sendable {
    case off = "OFF"
    case observe = "OBSERVE"
    case enforce = "ENFORCE"
}

public enum CalliopeiaAuditEffort: String, Codable, Sendable {
    case standard = "STANDARD"
    case maximum = "MAX"
}

public enum CalliopeiaAuditStrategy: String, Codable, Sendable {
    case combined = "COMBINED"
    case atomicBatch = "ATOMIC_BATCH"
}

public struct CalliopeiaAudioJobRequest: Sendable {
    public var processingProfileID: String?
    public var extractionEffort: CalliopeiaExtractionEffort?
    public var auditMode: CalliopeiaAuditMode?
    public var auditEffort: CalliopeiaAuditEffort?
    public var auditStrategy: CalliopeiaAuditStrategy?
    public var auditBatchSize: Int?
    public var promptText: String?
    public var promptTemplateID: String?
    public var promptTitle: String?
    public var responseMode: CalliopeiaResponseMode
    public var webhookEndpointID: String?
    public var idempotencyKey: String
    public var userID: String?
    public var shopID: String?
    public var customerID: String?
    public var expectedAddressee: String?
    public var karteID: String?
    public var summaryID: String?
    public var passthrough: JSONValue?

    public init(
        idempotencyKey: String = UUID().uuidString,
        processingProfileID: String? = nil,
        extractionEffort: CalliopeiaExtractionEffort? = nil,
        auditMode: CalliopeiaAuditMode? = nil,
        auditEffort: CalliopeiaAuditEffort? = nil,
        auditStrategy: CalliopeiaAuditStrategy? = nil,
        auditBatchSize: Int? = nil,
        promptText: String? = nil,
        promptTemplateID: String? = nil,
        promptTitle: String? = nil,
        responseMode: CalliopeiaResponseMode = .async,
        webhookEndpointID: String? = nil,
        userID: String? = nil,
        shopID: String? = nil,
        customerID: String? = nil,
        expectedAddressee: String? = nil,
        karteID: String? = nil,
        summaryID: String? = nil,
        passthrough: JSONValue? = nil
    ) {
        self.idempotencyKey = idempotencyKey
        self.processingProfileID = processingProfileID
        self.extractionEffort = extractionEffort
        self.auditMode = auditMode
        self.auditEffort = auditEffort
        self.auditStrategy = auditStrategy
        self.auditBatchSize = auditBatchSize
        self.promptText = promptText
        self.promptTemplateID = promptTemplateID
        self.promptTitle = promptTitle
        self.responseMode = responseMode
        self.webhookEndpointID = webhookEndpointID
        self.userID = userID
        self.shopID = shopID
        self.customerID = customerID
        self.expectedAddressee = expectedAddressee
        self.karteID = karteID
        self.summaryID = summaryID
        self.passthrough = passthrough
    }
}

public struct CalliopeiaUploadTicket: Codable, Equatable, Sendable {
    public let objectKey: String
    public let uploadURL: URL
    public let method: String
    public let contentType: String
    public let expiresAt: String?

    enum CodingKeys: String, CodingKey {
        case objectKey
        case uploadURL = "uploadUrl"
        case method
        case contentType
        case expiresAt
    }
}

public struct CalliopeiaAcceptedJob: Codable, Equatable, Sendable {
    public let id: String
    public let status: String
    public let externalSummaryID: String?
    public let externalIdempotencyKey: String?
    public let externalResponseMode: String?
    public let externalCallbackStatus: String?

    enum CodingKeys: String, CodingKey {
        case id, status
        case externalSummaryID = "externalSummaryId"
        case externalIdempotencyKey
        case externalResponseMode
        case externalCallbackStatus
    }
}

public struct CalliopeiaJobSubmission: Codable, Equatable, Sendable {
    public let message: String?
    public let idempotentReplay: Bool
    public let subscriptionToken: String?
    public let statusURL: URL?
    public let job: CalliopeiaAcceptedJob

    enum CodingKeys: String, CodingKey {
        case message, idempotentReplay, subscriptionToken, job
        case statusURL = "statusUrl"
    }
}

public struct CalliopeiaJobSnapshot: Codable, Equatable, Sendable {
    public let id: String
    public let status: String
    public let operation: String?
    public let inputKind: String?
    public let fileName: String?
    public let audioSeconds: Double?
    public let responseText: String?
    public let responseJSON: JSONValue?
    public let extractionEffort: String?
    public let auditMode: String?
    public let auditEffort: String?
    public let auditStrategy: String?
    public let auditBatchSize: Int?
    public let auditState: String?
    public let auditJSON: JSONValue?
    public let deliveryBlockedReason: String?
    public let errorMessage: String?
    public let costUSD: Double?
    public let costJPY: Double?
    public let createdAt: String?
    public let updatedAt: String?
    public let completedAt: String?
    public let externalShopID: String?
    public let externalCustomerID: String?
    public let externalKarteID: String?
    public let externalSummaryID: String?
    public let passthrough: JSONValue?

    enum CodingKeys: String, CodingKey {
        case id, status, operation, inputKind, fileName, audioSeconds, responseText
        case extractionEffort, auditEffort, auditStrategy, auditBatchSize
        case deliveryBlockedReason, errorMessage, createdAt, updatedAt, completedAt
        case responseJSON = "responseJson"
        case auditMode = "transcriptSupportAuditMode"
        case auditState = "transcriptSupportAuditState"
        case auditJSON = "transcriptSupportAuditJson"
        case costUSD = "costUsd"
        case costJPY = "costJpy"
        case externalShopID = "externalShopId"
        case externalCustomerID = "externalCustomerId"
        case externalKarteID = "externalKarteId"
        case externalSummaryID = "externalSummaryId"
        case passthrough = "externalPassthrough"
    }
}

public struct CalliopeiaPullJob: Codable, Equatable, Sendable {
    public let version: String
    public let jobID: String
    public let status: String
    public let shopID: String?
    public let customerID: String?
    public let karteID: String?
    public let summaryID: String?
    public let passthrough: JSONValue?
    public let error: JSONValue?
    public let delivery: JSONValue
    public let createdAt: String?
    public let updatedAt: String?
    public let completedAt: String?
    public let result: JSONValue?

    enum CodingKeys: String, CodingKey {
        case version, status, passthrough, error, delivery, createdAt, updatedAt, completedAt, result
        case jobID = "jobId"
        case shopID = "shopId"
        case customerID = "customerId"
        case karteID = "karteId"
        case summaryID = "summaryId"
    }
}

public struct CalliopeiaPullJobResponse: Codable, Equatable, Sendable {
    public let success: Bool
    public let job: CalliopeiaPullJob
    public let requestID: String

    enum CodingKeys: String, CodingKey {
        case success, job
        case requestID = "requestId"
    }
}

public enum CalliopeiaSDKError: Error, Equatable, Sendable {
    case invalidRequest(String)
    case invalidResponse(String)
    case service(String)
    case http(statusCode: Int, body: String?)
    case missingPullAPIBaseURL
}

extension CalliopeiaSDKError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .invalidRequest(let message): message
        case .invalidResponse(let message): message
        case .service(let message): message
        case .http(let statusCode, let body):
            body.map { "Calliopeia returned HTTP \(statusCode): \($0)" }
                ?? "Calliopeia returned HTTP \(statusCode)"
        case .missingPullAPIBaseURL:
            "pullAPIBaseURL is required to retrieve job results"
        }
    }
}
